test/typescript.ts

```diff
@@ -1,33 +1,68 @@
-import {Store, Middleware} from 'redux';
-import thunk, {ThunkAction} from '../index.d.ts';
+import { createStore, applyMiddleware } from 'redux';
+import thunk, { ThunkAction, ThunkMiddleware } from '../index';
 
+type State = {
+  foo: string;
+};
 
-declare const store: Store<{foo: string}>;
+type Actions = { type: 'FOO' } | { type: 'BAR', result: number };
 
-store.dispatch(dispatch => {
-  dispatch({type: 'FOO'});
-});
+type ThunkResult<R> = ThunkAction<R, State, undefined, Actions>;
 
-store.dispatch((dispatch, getState) => {
-  const state = getState();
+const initialState: State = {
+  foo: 'foo'
+};
 
-  const foo: string = state.foo;
-});
+function fakeReducer(state: State = initialState, action: Actions): State {
+  return state;
+}
 
-const middleware: Middleware = thunk.withExtraArgument('bar');
+const store = createStore(fakeReducer, applyMiddleware(thunk as ThunkMiddleware<State, Actions>));
 
-store.dispatch((dispatch, getState, extraArg) => {
-  console.log(extraArg);
+store.dispatch(dispatch => {
+  dispatch({ type: 'FOO' });
+  // typings:expect-error
+  dispatch({ type: 'BAR' })
+  dispatch({ type: 'BAR', result: 5 })
+  // typings:expect-error
+  store.dispatch({ type: 'BAZ'});
 });
 
-const thunkAction: ThunkAction<void, {foo: string}, {bar: number}> =
-  (dispatch, getState, extraArg) => {
-    const foo: string = getState().foo;
-    const bar: number = extraArg.bar;
-
-    dispatch({type: 'FOO'});
+function testGetState(): ThunkResult<void> {
+  return (dispatch, getState) => {
+    const state = getState();
+    const foo: string = state.foo;
+    dispatch({ type: 'FOO' });
+    // typings:expect-error
+    dispatch({ type: 'BAR'});
+    dispatch({ type: 'BAR', result: 5 });
+    // typings:expect-error
+    dispatch({ type: 'BAZ'});
+    // Can dispatch another thunk action
+    dispatch(testGetState());
```

Comment:
Won't this cause a stack overflow?  I think second simple method that returned a `ThunkResult` would solve this

```ts
function testGetState(): ThunkResult<void> {
  return (dispatch, getState) => {
    const state = getState();
    const foo: string = state.foo;
    dispatch({ type: 'FOO' });
    // typings:expect-error
    dispatch({ type: 'BAR'});
    dispatch({ type: 'BAR', result: 5 });
    // typings:expect-error
    dispatch({ type: 'BAZ'});
    // Can dispatch another thunk action
    dispatch(getStringThunk()); // <-- Updated this method call
  };
}

function getStringThunk(): ThunkResult<string> {
    return (dispatch, getState) => {
        return 'foobar';
    };
}
```

Reply:
Yes if you run it. but this file goes through type checking only, wont be executed by test runner.

Reply:
But don't you think this serves as a nice example file that someone might try running?

Reply:
I think you are right. We could change it to another thunk action.